File: src/levels/L03-builder/ts/decorator.ts
class Order {
  // region: Order#initialize
  constructor(private size: string, private temp: string) {}
  // endregion
}

// Los extras se apilan como envolturas con nombre: ya no se cruzan.
// region: WithMilk
class WithMilk {
  constructor(private inner: Order, private milk: string) {}
}
// endregion

// region: WithSyrup
class WithSyrup {
  constructor(private inner: Order | WithMilk, private syrup: string) {}
}
// endregion

class WithIce {
  constructor(private inner: Order) {}
  isIced() { return true }
}

class OrderTaker {
  constructor(private cashier: Cashier) {}

  // …pero nadie revisa el pedido completo: caliente + hielo pasa y explota en cocina.
  // region: OrderTaker#take
  take(request: OrderRequest) {
    const order = new WithSyrup(new WithMilk(new Order('grande', 'caliente'), 'avena'), 'vainilla')
    this.cashier.charge(order)
  }
  // endregion
}

File: src/levels/L03-builder/decorator.rb
require "delegate"

class Order
  # region: Order#initialize
  def initialize(size, temp) = (@size, @temp = size, temp)
  # endregion
end

# Los extras se apilan como envolturas con nombre: ya no se cruzan.
# region: WithMilk
class WithMilk < SimpleDelegator
  def initialize(order, milk) = (super(order); @milk = milk)
end
# endregion

# region: WithSyrup
class WithSyrup < SimpleDelegator
  def initialize(order, syrup) = (super(order); @syrup = syrup)
end
# endregion

class OrderTaker
  def initialize(cashier) = @cashier = cashier

  # …pero nadie revisa el pedido completo: caliente + hielo pasa y explota en cocina.
  # region: OrderTaker#take
  def take(request)
    order = WithSyrup.new(WithMilk.new(Order.new(:grande, :caliente), :avena), :vainilla)
    @cashier.charge(order)
  end
  # endregion
end
